packages/typescript/src/core/trace_manager.ts

```
@@ -0,0 +1,144 @@
+import { LiveSpan, Span } from './entities/span';
+import { TraceInfo } from './entities/trace_info';
+import { Trace } from './entities/trace';
+import { TraceData } from './entities/trace_data';
+import { SpanAttributeKey } from './constants';
+
+/**
+ * Internal representation to keep the state of a trace.
+ * Uses a Map<string, LiveSpan> instead of TraceData to allow access by span_id.
+ */
+class _Trace {
+  info: TraceInfo;
+  spanDict: Map<string, LiveSpan>;
+
+  constructor(info: TraceInfo) {
+    this.info = info;
+    this.spanDict = new Map<string, LiveSpan>();
+  }
+
+  /**
+   * Convert the internal trace representation to an MLflow Trace object
+   */
+  toMlflowTrace(): Trace {
+    const traceData = new TraceData();
+
+    // Convert LiveSpan, mutable objects, into immutable Span objects before persisting
+    for (const span of this.spanDict.values()) {
+      traceData.spans.push(span as Span);
+    }
+
+    const root_span = traceData.spans.find((span) => span.parentId == null);
```

Shall we throw an exception if `root_span` is not found?

---

When `toMlflowTrace` is executed, it is guaranteed that the trace has some root span (otherwise exporter will skip it). We can add invalid date exception to be super safe, but I think it's not necessary since we haven't faced it in python sdk.